File: src/2024/day1.rs
#[path = "../utils.rs"]
mod utils;
type Int = i32;

fn main() {
    // https://adventofcode.com/2024/day/1
    println!("Day 1, 2024: Historian Hysteria");
    let _m = utils::Monitor::start();

    let p1 = part1(get_input());
    println!("  Part 1: {}", p1);
    assert_eq!(2769675, p1);

    let p2 = part2(get_input());
    println!("  Part 2: {}", p2);
    assert_eq!(24643097, p2);
}

fn part1(input: &'static str) -> Int {
    let mut list_a: Vec<Int> = vec![];
    let mut list_b: Vec<Int> = vec![];

    for line in input.lines() {
        let mut parts = line.split_whitespace();
        list_a.push(parts.next().unwrap().parse().unwrap());
        list_b.push(parts.next().unwrap().parse().unwrap());
    }

    list_a.sort();
    list_b.sort();

    list_a
        .iter()
        .zip(list_b.iter())
        .map(|(&a, &b)| (a - b).abs())
        .sum()
}

fn part2(input: &'static str) -> Int {
    let mut list_a: Vec<Int> = vec![];
    let mut list_b: Vec<Int> = vec![];

    for line in input.lines() {
        let mut parts = line.split_whitespace();
        list_a.push(parts.next().unwrap().parse().unwrap());
        list_b.push(parts.next().unwrap().parse().unwrap());
    }

    list_a
        .iter()
        .map(|&a| a * list_b.iter().filter(|&&b| a == b).count() as Int)
        .sum()
}

fn get_input() -> &'static str {
    include_str!("../../inputs/2024/day1.txt").trim_end()
}

#[cfg(test)]
mod tests {
    fn get_example() -> &'static str {
        indoc::indoc! {"
            3   4
            4   3
            2   5
            1   3
            3   9
            3   3
        "}
    }

    #[test]
    fn part1() {
        assert_eq!(11, super::part1(get_example()));
    }

    #[test]
    fn part2() {
        assert_eq!(31, super::part2(get_example()));
    }
}

File: src/2024/day1.zig
const std = @import("std");
const utils = @import("utils");

pub fn main() !void {
    // https://adventofcode.com/2024/day/1
    std.debug.print("Day 1, 2024: Historian Hysteria\n", .{});

    const m = utils.Monitor.init();
    defer m.deinit();
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    const input = comptime utils.trim_input(@embedFile("input"));

    const p1 = try part1(input, allocator);
    std.debug.print("  Part 1: {}\n", .{p1});
    std.debug.assert(2769675 == p1);

    const p2 = try part2(input, allocator);
    std.debug.print("  Part 2: {}\n", .{p2});
    std.debug.assert(24643097 == p2);
}

const Lists = struct {
    a: std.ArrayList(i32),
    b: std.ArrayList(i32),

    pub fn init(ally: std.mem.Allocator) Lists {
        return Lists{
            .a = std.ArrayList(i32).init(ally),
            .b = std.ArrayList(i32).init(ally),
        };
    }

    pub fn parse(self: *@This(), input: []const u8) !void {
        var it = std.mem.tokenizeAny(u8, input, "\r\n");
        while (it.next()) |line| {
            var line_it = std.mem.tokenizeScalar(u8, line, ' ');
            const a = line_it.next().?;
            const b = line_it.next().?;
            try self.a.append(try std.fmt.parseInt(i32, a, 10));
            try self.b.append(try std.fmt.parseInt(i32, b, 10));
        }
    }

    pub fn deinit(self: *@This()) void {
        self.a.deinit();
        self.b.deinit();
    }
};

fn part1(input: []const u8, ally: std.mem.Allocator) !i32 {
    var lists = Lists.init(ally);
    defer lists.deinit();
    try lists.parse(input);

    std.mem.sort(i32, lists.a.items, {}, std.sort.asc(i32));
    std.mem.sort(i32, lists.b.items, {}, std.sort.asc(i32));

    var sum: i32 = 0;
    for (lists.a.items, lists.b.items) |a, b| {
        sum += @max(a, b) - @min(a, b);
    }

    return sum;
}

test "part 1" {
    const example =
        \\3   4
        \\4   3
        \\2   5
        \\1   3
        \\3   9
        \\3   3
    ;
    try std.testing.expectEqual(11, try part1(example, std.testing.allocator));
}

fn part2(input: []const u8, ally: std.mem.Allocator) !i32 {
    var lists = Lists.init(ally);
    defer lists.deinit();
    try lists.parse(input);

    var sum: i32 = 0;
    for (lists.a.items) |a| {
        var count: i32 = 0;
        for (lists.b.items) |b| {
            if (a == b) {
                count += 1;
            }
        }

        sum += a * count;
    }

    return sum;
}

test "part 2" {
    const example =
        \\3   4
        \\4   3
        \\2   5
        \\1   3
        \\3   9
        \\3   3
    ;
    try std.testing.expectEqual(31, try part2(example, std.testing.allocator));
}
